drivers/flash/flash_stm32_ospi.c

```diff
@@ -242,24 +245,79 @@ static OSPI_RegularCmdTypeDef ospi_prepare_cmd(uint8_t transfer_mode, uint8_t tr
 	return cmd_tmp;
 }
 
+/*
+ * Read Serial Flash ID :
+ * perform a read access over SPI bus for read Identification (DataMode is already set)
+ * and compare to the jedec-id from the DTYS table exists
+ */
+static int ospi_read_jedec_id(const struct device *dev,  uint8_t *id)
+{
+#if DT_NODE_HAS_PROP(DT_INST(0, st_stm32_ospi_nor), jedec_id)
+	const struct flash_stm32_ospi_config *dev_cfg = dev->config;
+#endif /* jedec_id */
+	struct flash_stm32_ospi_data *dev_data = dev->data;
+	uint8_t data[JESD216_READ_ID_LEN]; /* data received from the octoflash */
+
+	/* This is a SPI/STR command to issue to the octoFlash device */
+	OSPI_RegularCmdTypeDef cmd = ospi_prepare_cmd(OSPI_SPI_MODE, OSPI_STR_TRANSFER);
+
+	cmd.Instruction = JESD216_CMD_READ_ID;
+	cmd.DummyCycles = 8U;
+	cmd.AddressSize = HAL_OSPI_ADDRESS_NONE;
+	cmd.NbData = JESD216_READ_ID_LEN; /* 3 bytes in the READ ID */
+
+	HAL_StatusTypeDef hal_ret;
+
+	hal_ret = HAL_OSPI_Command(&dev_data->hospi, &cmd, HAL_OSPI_TIMEOUT_DEFAULT_VALUE);
+
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("%d: Failed to send OSPI instruction", hal_ret);
+		return -EIO;
+	}
+
+	hal_ret = HAL_OSPI_Receive(&dev_data->hospi, data, HAL_OSPI_TIMEOUT_DEFAULT_VALUE);
+	if (hal_ret != HAL_OK) {
+		LOG_ERR("%d: Failed to read data", hal_ret);
+		return -EIO;
+	}
+
+	/*
+	 * ReadID data received from the octoflash is now compared
+	 * to the jedec-id values from the DTS (if defined)
+	 */
+#if DT_NODE_HAS_PROP(DT_INST(0, st_stm32_ospi_nor), jedec_id)
+	if ((data[1] != dev_cfg->jedec_id[0]) &&  /* manufacturer ID */
+		(data[2] != dev_cfg->jedec_id[1]) &&  /* memory Type */
+		(data[0] != dev_cfg->jedec_id[2])) /* memory Density */ {
+
+		LOG_ERR("Device id %02x %02x %02x does not match config %02x %02x %02x",
+			data[1], data[2], data[0],
+			dev_cfg->jedec_id[0], dev_cfg->jedec_id[1], dev_cfg->jedec_id[2]);
+		return -EINVAL;
+	}
+	LOG_INF("Manuf ID = %02x   Memory Type = %02x   Memory Density = %02x",
+		data[0], data[1], data[2]);
+#endif /* jedec_id */
+	dev_data->cmd_status = 0;
+	id = &data[0];
```

Comment:
You are assigning `id` pointer to a local variable:
```
	id = &data[0];
```
The `id` correctness is not guaranteed upon exit from the `ospi_read_jedec_id`.

A fix would be to expect the `id` to point to an array of size `JESD216_READ_ID_LEN` outside of the function.